swift/StableDiffusion/pipeline/StableDiffusionXLPipeline.swift

```diff
@@ -160,21 +187,53 @@ public struct StableDiffusionXLPipeline: StableDiffusionPipelineProtocol {
         if reduceMemory {
             encoder?.unloadResources()
         }
+
         let timestepStrength: Float? = config.mode == .imageToImage ? config.strength : nil
-        
-        // De-noising loop
+
+        // Store current model
+        var unetModel = self.unet
+        var currentInput = baseInput ?? refinerInput
+
+        var unetHiddenStates = currentInput?.hiddenStates
+        var unetPooledStates = currentInput?.pooledStates
+        var unetGeometryConditioning = currentInput?.geometryConditioning
+
         let timeSteps: [Int] = scheduler[0].calculateTimesteps(strength: timestepStrength)
-        for (step,t) in timeSteps.enumerated() {
 
+        // Calculate which step to swap to refiner
+        let refinerFraction: Float = config.strength
+        let refinerStart = Int(Float(timeSteps.count) * refinerFraction)
+
+        // De-noising loop
+        for (step,t) in timeSteps.enumerated() {
             // Expand the latents for classifier-free guidance
             // and input to the Unet noise prediction model
             let latentUnetInput = latents.map {
                 MLShapedArray<Float32>(concatenating: [$0, $0], alongAxis: 0)
             }
-                        
+
+            // Switch to refiner if specified
+            if unetRefiner != nil, step == refinerStart {
+                if reduceMemory {
+                    self.unet.unloadResources()
```

Comment:
Good point, also since `reduceMemory` is computed it would never be true anyway. I'd just be cautious of how much loading the refiner from disk would slow down the overall process. What do you think about a config that is user configurable so that they can choose to keep them both in memory or not, similar to how some of the web uis use `--lowvram` (unload all models when not in use) and `--medvram` (only unload the unets)?

Reply:
IMO, this might get complex quickly.. @pcuenca has a point that base+refiner is usually >10GB which is higher than the available RAM under normal system load in a significant fraction of the Mac(Book)s today